File: src/main/java/toby/handler/Handler.java
package toby.handler;

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.entities.channel.middleman.AudioChannel;
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel;
import net.dv8tion.jda.api.entities.emoji.Emoji;
import net.dv8tion.jda.api.events.guild.GuildAvailableEvent;
import net.dv8tion.jda.api.events.guild.GuildJoinEvent;
import net.dv8tion.jda.api.events.guild.GuildReadyEvent;
import net.dv8tion.jda.api.events.guild.voice.GuildVoiceUpdateEvent;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
import net.dv8tion.jda.api.events.interaction.component.StringSelectInteractionEvent;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.events.session.ReadyEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.dv8tion.jda.api.managers.AudioManager;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Configurable;
import org.springframework.stereotype.Service;
import toby.BotMain;
import toby.emote.Emotes;
import toby.helpers.HttpHelper;
import toby.jpa.dto.ConfigDto;
import toby.jpa.dto.UserDto;
import toby.jpa.service.*;
import toby.lavaplayer.GuildMusicManager;
import toby.lavaplayer.PlayerManager;
import toby.managers.CommandManager;

import javax.annotation.Nonnull;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import static toby.command.commands.fetch.DnDCommand.doLookUpAndReply;
import static toby.helpers.MusicPlayerHelper.playUserIntroWithChannel;
import static toby.helpers.UserDtoHelper.calculateUserDto;

@Service
@Configurable
public class Handler extends ListenerAdapter {

    private final IConfigService configService;
    private final IBrotherService brotherService;
    private final IUserService userService;
    private final IMusicFileService musicFileService;
    private final IExcuseService excuseService;
    private static final Logger LOGGER = LoggerFactory.getLogger(Handler.class);
    private final CommandManager manager;

    @Autowired
    public Handler(IConfigService configService, IBrotherService brotherService, IUserService userService, IMusicFileService musicFileService, IExcuseService excuseService) {
        manager = new CommandManager(configService, brotherService, userService, musicFileService, excuseService);
        this.configService = configService;
        this.brotherService = brotherService;
        this.userService = userService;
        this.musicFileService = musicFileService;
        this.excuseService = excuseService;
    }


    @Override
    public void onReady(@Nonnull ReadyEvent event) {
        LOGGER.info("{} is ready", event.getJDA().getSelfUser().getAsTag());
    }

    @Override
    public void onMessageReceived(@Nonnull MessageReceivedEvent event) {
        User user = event.getAuthor();

        if (user.isBot() || event.isWebhookMessage()) {
            return;
        }

        //Event specific information
        User author = event.getAuthor();                //The user that sent the message
        Message message = event.getMessage();           //The message that was received.
        MessageChannel channel = event.getChannel();    //This is the MessageChannel that the message was sent to.
        //  This could be a TextChannel, PrivateChannel, or Group!

        String msg = message.getContentDisplay();              //This returns a human readable version of the Message. Similar to
        // what you would see in the client.

        boolean bot = author.isBot();                    //This boolean is useful to determine if the User that
        // sent the Message is a BOT or not!


        if (bot) {
            return;
        }
        Guild guild = event.getGuild();             //The Guild that this message was sent in. (note, in the API, Guilds are Servers)
        TextChannel textChannel = event.getChannel().asTextChannel(); //The TextChannel that this message was sent to.
        Member member = event.getMember();          //This Member that sent the message. Contains Guild specific information about the User!

        String name;
        if (message.isWebhookMessage()) {
            name = author.getName();                //If this is a Webhook message, then there is no Member associated
        }                                           // with the User, thus we default to the author for name.
        else {
            name = member.getEffectiveName();       //This will either use the Member's nickname if they have one,
        }                                           // otherwise it will default to their username. (User#getName())

        JDA jda = guild.getJDA();
        Emoji tobyEmote = jda.getEmojiById(Emotes.TOBY);
        Emoji jessEmote = jda.getEmojiById(Emotes.JESS);

        messageContainsRespond(message, channel, name, tobyEmote, jessEmote);

    }

    @Override
    public void onSlashCommandInteraction(@NotNull SlashCommandInteractionEvent event) {
        User user = event.getUser();
        if (user.isBot()) {
            return;
        }
        manager.handle(event);
    }

    @Override
    public void onButtonInteraction(ButtonInteractionEvent event) {
        User user = event.getUser();
        if (user.isBot()) {
            return;
        }
        manager.handle(event);
    }

    @Override
    public void onGuildReady(@NotNull GuildReadyEvent event) {
        event.getGuild().updateCommands().addCommands(manager.getAllSlashCommands()).queue();
    }

    @Override
    public void onGuildJoin(@NotNull GuildJoinEvent event) {
        event.getGuild().updateCommands().addCommands(manager.getAllSlashCommands()).queue();
    }

    @Override
    public void onGuildAvailable(@NotNull GuildAvailableEvent event) {
        event.getGuild().updateCommands().addCommands(manager.getAllSlashCommands()).queue();
    }

    private void messageContainsRespond(Message message, MessageChannel channel, String name, Emoji tobyEmote, Emoji jessEmote) {
        String messageStringLowercase = message.getContentRaw().toLowerCase();
        if (messageStringLowercase.contains("toby") || messageStringLowercase.contains("tobs")) {
            message.addReaction(tobyEmote).queue();
            channel.sendMessageFormat("%s... that's not my name %s", name, tobyEmote).queue();
        }

        if (messageStringLowercase.trim().contains("sigh")) {
            channel.sendMessageFormat("Hey %s, what's up champ?", name).queue();
            channel.sendMessageFormat("%s", jessEmote).queue();
        }

        if (messageStringLowercase.contains("yeah")) {
            channel.sendMessage("YEAH????").queue();
        }

        if (message.getMentions().isMentioned(BotMain.getJda().getSelfUser())) {
            channel.sendMessage("Don't talk to me").queue();
        }

        if (messageStringLowercase.contains("covid") || messageStringLowercase.contains("corona")) {
            channel.sendMessage("""
                    It is the 2nd millennium, for more than two years humanity has sat immobile on it's fat arse whilst COVID roams the planet. They are the masters of Netflix by the will of the settee, and masters of ordering Chinese takeaway through the might of their wallets. They are fattening imbeciles imbued with energy from last nights curry. They are the isolated ones for whom more than a million people wear masks every day.

                    Yet, even in their quarantined state, they continue to make everyone's lives miserable. Arsehole scalpers plunder the Internet for the latest ps5 stock, hoarding until they can raise the prices further still. Greatest among these cretins are the anti-vaxxers, complete morons, idiots with IQs less than a goat. Their fools in arms are endless: flat earthers and countless moon landing deniers, the stubborn Facebook politicians and the Karen's of every shopping centre to name only a few. And with all this nonsense they won't shut up about 5G, Immigrants, muh freedoms... and far, far worse.

                    To be a man (or woman) in such times is to be one amongst almost 7 billion. It is to live in the stupidest and most irritating regime imaginable. These are the memes of these times. Forget the power of technology and science, for so much will be denied, never to be acknowledged. Forget the promise of immunity and vaccinations, for in our grim dark daily lives, there is only COVID. There is no freedom on our streets, only an eternity of mask mandates and sanitizer, and the coughing of the sick.""").queue();
        }
    }


    @Override
    public void onGuildVoiceUpdate(@NotNull GuildVoiceUpdateEvent event) {
        if (event.getChannelJoined() != null) {
            onGuildVoiceJoin(event);
        }
        if (event.getChannelLeft() != null) {
            onGuildVoiceLeave(event);
        }
        if (event.getChannelJoined() == null && event.getChannelLeft() == null) {
            onGuildVoiceMove(event);
        }
    }

    //Auto joining voice channel when it becomes occupied and an audio connection doesn't already exist on the server, then play the associated user's intro song

    public void onGuildVoiceJoin(GuildVoiceUpdateEvent event) {
        Guild guild = event.getGuild();
        AudioManager audioManager = guild.getAudioManager();
        String volumePropertyName = ConfigDto.Configurations.VOLUME.getConfigValue();
        ConfigDto databaseVolumeConfig = configService.getConfigByName(volumePropertyName, event.getGuild().getId());
        ConfigDto deleteDelayConfig = configService.getConfigByName(ConfigDto.Configurations.DELETE_DELAY.getConfigValue(), event.getGuild().getId());

        int defaultVolume = databaseVolumeConfig != null ? Integer.parseInt(databaseVolumeConfig.getValue()) : 100;
        List<Member> nonBotConnectedMembers = event.getChannelJoined().getMembers().stream().filter(member -> !member.getUser().isBot()).toList();
        AudioPlayer audioPlayer = PlayerManager.getInstance().getMusicManager(guild).getAudioPlayer();
        if (!nonBotConnectedMembers.isEmpty() && !audioManager.isConnected()) {
            audioPlayer.setVolume(defaultVolume);
            audioManager.openAudioConnection(event.getChannelJoined());
        }
        Member member = event.getMember();
        long discordId = member.getUser().getIdLong();
        long guildId = member.getGuild().getIdLong();
        UserDto requestingUserDto = calculateUserDto(guildId, discordId, Objects.requireNonNull(event.getMember()).isOwner(), userService, defaultVolume);

        if (Objects.equals(audioManager.getConnectedChannel(), event.getChannelJoined())) {
            playUserIntroWithChannel(requestingUserDto, guild, guild.getDefaultChannel().asTextChannel(), Integer.parseInt(deleteDelayConfig.getValue()), 0L);
        }
    }

    public void onGuildVoiceMove(GuildVoiceUpdateEvent event) {
        Guild guild = event.getGuild();
        AudioManager audioManager = guild.getAudioManager();
        String volumePropertyName = ConfigDto.Configurations.VOLUME.getConfigValue();
        ConfigDto databaseConfig = configService.getConfigByName(volumePropertyName, event.getGuild().getId());
        int defaultVolume = databaseConfig != null ? Integer.parseInt(databaseConfig.getValue()) : 100;
        List<Member> nonBotConnectedMembersInOldChannel = event.getChannelLeft().getMembers().stream().filter(member -> !member.getUser().isBot()).collect(Collectors.toList());
        closeAudioManagerIfChannelEmpty(guild, audioManager, defaultVolume, nonBotConnectedMembersInOldChannel, event.getChannelLeft());
        List<Member> nonBotConnectedMembers = event.getChannelJoined().getMembers().stream().filter(member -> !member.getUser().isBot()).toList();
        if (!nonBotConnectedMembers.isEmpty() && !audioManager.isConnected()) {
            PlayerManager.getInstance().getMusicManager(guild).getAudioPlayer().setVolume(defaultVolume);
            audioManager.openAudioConnection(event.getChannelJoined());
        }
        deleteTemporaryChannelIfEmpty(nonBotConnectedMembers.isEmpty(), event.getChannelLeft());
    }


    //Auto leaving voice channel when it becomes empty
    public void onGuildVoiceLeave(GuildVoiceUpdateEvent event) {
        Guild guild = event.getGuild();
        AudioManager audioManager = guild.getAudioManager();
        String volumePropertyName = ConfigDto.Configurations.VOLUME.getConfigValue();
        ConfigDto databaseConfig = configService.getConfigByName(volumePropertyName, event.getGuild().getId());
        int defaultVolume = databaseConfig != null ? Integer.parseInt(databaseConfig.getValue()) : 100;
        List<Member> nonBotConnectedMembers = event.getChannelLeft().getMembers().stream().filter(member -> !member.getUser().isBot()).collect(Collectors.toList());
        closeAudioManagerIfChannelEmpty(guild, audioManager, defaultVolume, nonBotConnectedMembers, event.getChannelLeft());
        deleteTemporaryChannelIfEmpty(nonBotConnectedMembers.isEmpty(), event.getChannelLeft());
    }

    private void closeAudioPlayer(Guild guild, AudioManager audioManager, int defaultVolume) {
        GuildMusicManager musicManager = PlayerManager.getInstance().getMusicManager(guild);
        musicManager.getScheduler().setLooping(false);
        musicManager.getScheduler().getQueue().clear();
        musicManager.getAudioPlayer().stopTrack();
        musicManager.getAudioPlayer().setVolume(defaultVolume);
        audioManager.closeAudioConnection();
    }

    private void closeAudioManagerIfChannelEmpty(Guild guild, AudioManager audioManager, int defaultVolume, List<Member> nonBotConnectedMembers, AudioChannel channelLeft) {
        if (Objects.equals(audioManager.getConnectedChannel(), channelLeft) && nonBotConnectedMembers.isEmpty()) {
            closeAudioPlayer(guild, audioManager, defaultVolume);
        }
    }

    private void deleteTemporaryChannelIfEmpty(boolean nonBotConnectedMembersEmpty, AudioChannel channelLeft) {
        //The autogenerated channels from the team command are "Team #", so this will delete them once they become empty
        if (channelLeft.getName().matches("(?i)team\\s[0-9]+") && nonBotConnectedMembersEmpty) {
            channelLeft.delete().queue();
        }
    }

    @Override
    public void onStringSelectInteraction(StringSelectInteractionEvent event) {
        event.deferReply().queue();
        if (event.getComponentId().equals("DnDspells")) {
            sendDndApiRequest(event, "spells");
        }
        if (event.getComponentId().equals("DnDconditions")) {
            sendDndApiRequest(event, "conditions");
        }
    }

    private void sendDndApiRequest(StringSelectInteractionEvent event, String spells) {
        String selectedValue = event.getValues().get(0); // Get the selected option
        ConfigDto deleteDelayConfig = configService.getConfigByName(ConfigDto.Configurations.DELETE_DELAY.getConfigValue(), event.getGuild().getId());
        event.getMessage().delete().queue();
        doLookUpAndReply(event.getHook(), spells, selectedValue, new HttpHelper(), Integer.valueOf(deleteDelayConfig.getValue()));
    }

}

File: src/main/kotlin/toby/handler/Handler.kt
package toby.handler

import mu.KotlinLogging
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.channel.concrete.VoiceChannel
import net.dv8tion.jda.api.entities.channel.middleman.AudioChannel
import net.dv8tion.jda.api.events.guild.GuildAvailableEvent
import net.dv8tion.jda.api.events.guild.GuildJoinEvent
import net.dv8tion.jda.api.events.guild.GuildReadyEvent
import net.dv8tion.jda.api.events.guild.voice.GuildVoiceUpdateEvent
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.StringSelectInteractionEvent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.events.session.ReadyEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.managers.AudioManager
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Configurable
import org.springframework.stereotype.Service
import toby.BotMain.Companion.jda
import toby.emote.Emotes
import toby.helpers.HttpHelper
import toby.helpers.MusicPlayerHelper.playUserIntro
import toby.helpers.UserDtoHelper.calculateUserDto
import toby.jpa.dto.ConfigDto
import toby.jpa.dto.UserDto
import toby.jpa.service.*
import toby.lavaplayer.PlayerManager
import toby.managers.ButtonManager
import toby.managers.CommandManager
import toby.managers.MenuManager
import java.util.*

@Service
@Configurable
class Handler @Autowired constructor(
    private val configService: IConfigService,
    brotherService: IBrotherService,
    private val userService: IUserService,
    musicFileService: IMusicFileService,
    excuseService: IExcuseService,
    httpHelper: HttpHelper,
    private val commandManager: CommandManager = CommandManager(
        configService,
        brotherService,
        userService,
        musicFileService,
        excuseService,
        httpHelper
    ),
    private val buttonManager: ButtonManager = ButtonManager(configService, userService, commandManager),
    private val menuManager: MenuManager = MenuManager(configService, httpHelper)
) : ListenerAdapter() {


    override fun onReady(event: ReadyEvent) {
        logger.info("${event.jda.selfUser.name} is ready")
        event.jda.guildCache.forEach { it.connectToMostPopulatedVoiceChannel() }
    }

    private fun Guild.connectToMostPopulatedVoiceChannel() {
        val mostPopulatedChannel = this.voiceChannels
            .filter { channel -> channel.members.any { !it.user.isBot } }
            .maxByOrNull { channel -> channel.members.count { !it.user.isBot } }

        if (mostPopulatedChannel != null && mostPopulatedChannel.members.count { !it.user.isBot } > 0) {
            mostPopulatedChannel.connectToVoiceChannel()
        } else {
            logger.info("No occupied voice channel to join found in guild: ${this.name}")
        }
    }

    private fun VoiceChannel.connectToVoiceChannel() {
        val guild = this.guild
        val audioManager = guild.audioManager
        if (!audioManager.isConnected) {
            audioManager.openAudioConnection(this)
            logger.info { "Connected to voice channel: ${this.name} in guild: ${guild.name}" }
        }
    }

    override fun onMessageReceived(event: MessageReceivedEvent) {
        val message = event.message
        val author = event.author
        val channel = event.channel
        val guild = event.guild
        val member = event.member

        if (author.isBot || event.isWebhookMessage) return

        val messageStringLowercase = message.contentRaw.lowercase(Locale.getDefault())

        when {
            messageStringLowercase.contains("toby") || messageStringLowercase.contains("tobs") -> {
                val tobyEmote = guild.jda.getEmojiById(Emotes.TOBY)
                channel.sendMessageFormat(
                    "%s... that's not my name %s",
                    member?.effectiveName ?: author.name,
                    tobyEmote
                ).queue()
                message.addReaction(tobyEmote!!).queue()
            }

            messageStringLowercase.trim() == "sigh" -> {
                val jessEmote = guild.jda.getEmojiById(Emotes.JESS)
                channel.sendMessageFormat("Hey %s, what's up champ?", member?.effectiveName ?: author.name, jessEmote).queue()
            }

            messageStringLowercase.contains("yeah") -> {
                channel.sendMessage("YEAH????").queue()
            }

            jda?.selfUser?.let { message.mentions.isMentioned(it) } == true -> {
                channel.sendMessage("Don't talk to me").queue()
            }
        }
    }

    override fun onSlashCommandInteraction(event: SlashCommandInteractionEvent) {
        if (!event.user.isBot) {
            commandManager.handle(event)
        }
    }

    override fun onButtonInteraction(event: ButtonInteractionEvent) {
        event.deferReply(true).queue()
        if (!event.user.isBot) {
            buttonManager.handle(event)
        }
    }

    override fun onGuildReady(event: GuildReadyEvent) {
        updateCommands(event.guild)
    }

    override fun onGuildJoin(event: GuildJoinEvent) {
        updateCommands(event.guild)
    }

    override fun onGuildAvailable(event: GuildAvailableEvent) {
        updateCommands(event.guild)
    }

    private fun updateCommands(guild: Guild) {
        guild.updateCommands().addCommands(commandManager.allSlashCommands).queue()
    }

    override fun onGuildVoiceUpdate(event: GuildVoiceUpdateEvent) {
        val guildId = event.guild.id

        when {
            event.channelJoined != null && event.channelLeft != null -> {
                logger.info("Voice move event triggered for guild $guildId from channel ${event.channelLeft} to channel ${event.channelJoined} for user ${event.member.effectiveName}")
                onGuildVoiceMove(event)
            }

            event.channelJoined != null -> {
                logger.info("Voice join event triggered for guild $guildId in channel ${event.channelJoined} for user ${event.member.effectiveName}")
                onGuildVoiceJoin(event)
            }

            event.channelLeft != null -> {
                logger.info("Voice leave event triggered for guild $guildId from channel ${event.channelLeft} for user ${event.member.effectiveName}")
                onGuildVoiceLeave(event)
            }
        }
    }

    private fun onGuildVoiceMove(event: GuildVoiceUpdateEvent) {
        val member = event.member
        if (member.user.jda.selfUser == jda?.selfUser && !event.guild.audioManager.isConnected) {
            val previousChannel = event.channelLeft?.asVoiceChannel()
            if (previousChannel != null) {
                rejoinPreviousChannel(event.guild, previousChannel)
            } else {
                logger.warn("Bot was moved from a channel but no previous channel found.")
            }
        }
    }

    private fun rejoinPreviousChannel(guild: Guild, channel: VoiceChannel) {
        runCatching {
            guild.audioManager.openAudioConnection(channel)
            logger.info("Rejoined previous channel '${channel.name}' on guild '${guild.id}'")
        }.onFailure {
            logger.error("Failed to rejoin channel '${channel.name}' on guild '${guild.id}': ${it.message}")
        }
    }

    private fun onGuildVoiceJoin(event: GuildVoiceUpdateEvent) {
        val guild = event.guild
        val audioManager = guild.audioManager
        val defaultVolume = getConfigValue(ConfigDto.Configurations.VOLUME.configValue, guild.id)
        val deleteDelayConfig =
            configService.getConfigByName(ConfigDto.Configurations.DELETE_DELAY.configValue, guild.id)

        val nonBotConnectedMembers = event.channelJoined?.members?.filter { !it.user.isBot } ?: emptyList()
        if (nonBotConnectedMembers.isNotEmpty() && !audioManager.isConnected) {
            PlayerManager.instance.getMusicManager(guild).audioPlayer.volume = defaultVolume
            audioManager.openAudioConnection(event.channelJoined)
        }

        if (audioManager.connectedChannel == event.channelJoined) {
            logger.info("Audiomanager channel ${audioManager.connectedChannel} and event joined channel ${event.channelJoined} are the same")
            setupAndPlayUserIntro(event.member, guild, defaultVolume, deleteDelayConfig)
        }
    }

    private fun setupAndPlayUserIntro(member: Member, guild: Guild, defaultVolume: Int, deleteDelayConfig: ConfigDto?) {
        val requestingUserDto = getRequestingUserDto(member, defaultVolume)
        playUserIntro(requestingUserDto, guild, deleteDelayConfig?.value?.toInt() ?: 0)
    }

    private fun onGuildVoiceLeave(event: GuildVoiceUpdateEvent) {
        val guild = event.guild
        val audioManager = guild.audioManager
        audioManager.checkAudioManagerToCloseConnectionOnEmptyChannel()
        deleteTemporaryChannelIfEmpty(event.channelLeft?.members?.none { !it.user.isBot } ?: true, event.channelLeft)
    }

    private fun AudioManager.checkAudioManagerToCloseConnectionOnEmptyChannel() {
        val connectedChannel = this.connectedChannel
        if (connectedChannel != null) {
            if (connectedChannel.members.none { !it.user.isBot }) {
                this.closeAudioConnection()
                logger.info("Audio connection closed on guild ${this.guild.id} due to empty channel.")
            }
        }
    }

    private fun deleteTemporaryChannelIfEmpty(nonBotConnectedMembersEmpty: Boolean, channelLeft: AudioChannel?) {
        if (channelLeft?.name?.matches("(?i)team\\s[0-9]+".toRegex()) == true && nonBotConnectedMembersEmpty) {
            logger.info("Deleting temporary channel: {}", channelLeft.name)
            channelLeft.delete().queue()
        }
    }

    private fun getConfigValue(configName: String, guildId: String, defaultValue: Int = 100): Int {
        val config = configService.getConfigByName(configName, guildId)
        return config?.value?.toInt() ?: defaultValue
    }

    private fun getRequestingUserDto(member: Member, defaultVolume: Int): UserDto {
        val discordId = member.idLong
        val guildId = member.guild.idLong
        return calculateUserDto(guildId, discordId, member.isOwner, userService, defaultVolume)
    }

    override fun onStringSelectInteraction(event: StringSelectInteractionEvent) {
        logger.info { "StringSelectInteractionEvent received on guild ${event.guild?.idLong}" }
        menuManager.handle(event)
    }

    companion object {
        private val logger = KotlinLogging.logger {}
    }
}
